File: annotatedcontigs2json/0.2/annotatedcontigs2json.py
#!/usr/bin/env python3

# (c) Christian Henke
# Licensed under Apache License 2.0
# https://www.apache.org/licenses/LICENSE-2.0

from Bio import SeqIO
import argparse
import json
import pysam
import numpy
from math import ceil
from multiprocessing import Pool, cpu_count

parser = argparse.ArgumentParser(description='''
Convert fasta to Elasticsearch json while including contig based annotation information and coverage
''')
parser.add_argument("--fasta", dest='input_fasta', required=True, help='fasta input file')
parser.add_argument("--virsorter-csv", dest='input_virsorter', required=True, help='VirSorter input file')
parser.add_argument('--sample-names', nargs='*', dest='names')
parser.add_argument('--sample-bam-files', nargs='*', dest='bams')
parser.add_argument("--json", dest='output_json', required=True, help='json output file')
parser.add_argument("--threads", type=int, dest='threads', help='number parallel threads for BAM file reader',
                    default=cpu_count())
args = parser.parse_args()

threads = args.threads

coverage_avg_span = 10
ones_val = int(coverage_avg_span / 2)

contigid_to_virsorter_phage = {}
contigid_to_virsorter_prophage = {}

coverage_map = {}


def parse_bam_for_ids(chunk, sample_name, bam_path):
  print('Parsing chunk of size %s' % len(chunk))
  coverage_map_chunk = {}
  sam_tmp = pysam.AlignmentFile(bam_path, "rb")
  for ref in chunk:
    ref_id = ref.split(' ', 1)[0]
    if ref_id not in coverage_map_chunk:
      coverage_map_chunk[ref_id] = {}
    try:
      cov_vals = [p.n for p in sam_tmp.pileup(ref)]
      if len(cov_vals) < coverage_avg_span:
        coverage_map_chunk[ref_id][sample_name] = []
      else:
        coverage_map_chunk[ref_id][sample_name] = numpy.convolve(cov_vals, numpy.ones(ones_val, ) / ones_val,
                                                                 mode='same')[
                                                  ones_val - 1:-ones_val:coverage_avg_span].astype(int).tolist()
    except ValueError:
      print('ERROR: Cannot find index file %s.bai' % bam_path)
      return False
    except:
      print('ERROR: Unknown error in thread.')
      return False
  print('Chunk done')
  return coverage_map_chunk


def merge_with_global_coverage_map(map_chunk):
  for ref_id, ref_obj in map_chunk.items():
    if ref_id in coverage_map:
      coverage_map[ref_id].update(ref_obj)
    else:
      coverage_map[ref_id] = ref_obj


if args.names and args.bams:
  named_bams = zip(args.names, args.bams)

  for name, bam in named_bams:
    print("For sample name '%s' parsing BAM file: %s" % (name, bam))
    sam = pysam.AlignmentFile(bam, "rb")
    total_ref_count = len(sam.references)
    print('File has %s refs' % total_ref_count)
    chunk_size = ceil(total_ref_count / threads)
    chunks = [sam.references[i:i + chunk_size] for i in range(0, len(sam.references), chunk_size)]
    print('Using %s threads' % threads)
    # real processes are used instead of threads to get around the CPython Global Interpreter Lock
    with Pool(processes=threads) as pool:
      results = [pool.apply_async(parse_bam_for_ids, args=(chunk, name, bam)) for chunk in chunks]
      for result in results:
        proc_result = result.get()
        if proc_result:
          merge_with_global_coverage_map(proc_result)
        else:
          exit(1)

with open(args.input_fasta, 'rU') as sourceFile, open(args.input_virsorter, 'r') as virsorterFile:
  phage_confidence = ''
  prophage_confidence = ''
  state = ''
  for line in virsorterFile:
    if line.strip() == '## 1 - Complete phage contigs - category 1 (sure)':
      phage_confidence = 'sure'
      state = 'phage'
    elif line.strip() == '## 2 - Complete phage contigs - category 2 (somewhat sure)':
      phage_confidence = 'somewhat sure'
      state = 'phage'
    elif line.strip() == '## 3 - Complete phage contigs - category 3 (not so sure)':
      phage_confidence = 'not so sure'
      state = 'phage'
    elif line.strip() == '## 4 - Prophages - category 1 (sure)':
      prophage_confidence = 'sure'
      state = 'prophage'
    elif line.strip() == '## 5 - Prophages - category 2 (somewhat sure)':
      prophage_confidence = 'somewhat sure'
      state = 'prophage'
    elif line.strip() == '## 6 - Prophages - category 3 (not so sure)':
      prophage_confidence = 'not so sure'
      state = 'prophage'
    else:
      if not line.startswith('#'):
        contigid = line.split(',')[0][len('VIRSorter_'):line.index('_flag=')]
        if state == 'phage':
          contigid_to_virsorter_phage[contigid] = phage_confidence
        elif state == 'prophage':
          contigid_to_virsorter_prophage[contigid] = prophage_confidence

  with open(args.output_json, 'w') as destFile:
    print('Writing output to %s' % args.output_json)
    for record in SeqIO.parse(sourceFile, "fasta"):
      id = str(record.id)
      header = {"index": {"_id": str(record.id)}}
      obj = {
        "id": str(record.id),
        "nucleotide": str(record.seq),
        "length": len(str(record.seq))
      }
      if args.names and args.bams and str(record.id) in coverage_map:
        obj['coverage'] = [{'name': k, 'values': v} for k, v in coverage_map[str(record.id)].items()]
      if id in contigid_to_virsorter_phage or id in contigid_to_virsorter_prophage:
        obj["virsorter"] = {}
        if id in contigid_to_virsorter_phage:
          obj["virsorter"]["phage"] = contigid_to_virsorter_phage[id]
        if id in contigid_to_virsorter_prophage:
          obj["virsorter"]["prophage"] = contigid_to_virsorter_prophage[id]
      destFile.write(json.dumps(header))
      destFile.write('\n')
      destFile.write(json.dumps(obj, separators=(',', ':')))
      destFile.write('\n')

File: annotatedcontigs2json/0.4.3/annotatedcontigs2json.go
/*
(c) 2019 Christian Henke
Licensed under Apache License 2.0
https://www.apache.org/licenses/LICENSE-2.0


Convert fasta to Elasticsearch JSON while including contig-based annotation information and coverage.
*/

package main

import (
	"bufio"
	"bytes"
	"compress/gzip"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"github.com/biogo/hts/bam"
	"github.com/biogo/hts/bgzf"
	"github.com/biogo/hts/bgzf/cache"
	"github.com/biogo/hts/sam"
	"github.com/shenwei356/bio/seq"
	"github.com/shenwei356/bio/seqio/fastx"
	"io"
	"log"
	"math"
	"os"
	"runtime"
	"runtime/pprof"
	"strconv"
	"strings"
	"sync"
	"time"
)

const CovAvgSpan int = 10

type Contig struct {
	ID         string      `json:"id"`
	Nucleotide string      `json:"nucleotide"`
	Length     int         `json:"length"`
	Coverage   []*Coverage `json:"coverage,omitempty"`
	Virsorter  *Virsorter  `json:"virsorter,omitempty"`
	Rpkm       []*Rpkm     `json:"rpkm,omitempty"`
}

type Coverage struct {
	Name   string `json:"name"`
	Values []int  `json:"values"`
}

type Virsorter struct {
	Phage    string `json:"phage,omitempty"`
	Prophage string `json:"prophage,omitempty"`
}

type Rpkm struct {
	Name  string  `json:"name"`
	Value float64 `json:"value"`
}

type Contigs map[string]*Contig

var (
	fastaFile       = flag.String("fasta", "", "FASTA input `file`")
	virSorterCSV    = flag.String("virsorter-csv", "", "VirSorter input `file`")
	sampleNames     = flag.String("sample-names", "", "Comma-separated `list` of sample names")
	sampleBAMFiles  = flag.String("sample-bam-files", "", "Comma-separated `list` of sorted BAM files")
	sampleRPKMFiles = flag.String("sample-rpkm-files", "", "Comma-separated `list` of BBMap .rpkm `files` (gzipped)")
	jsonOutputFile  = flag.String("json", "", "JSON output `file`")
	skipCoverage    = flag.Bool("skip-coverage", false, "Skip detailed coverage calculation (BAM) entirely")
	reduceRAMUsage  = flag.Bool("reduce-ram-usage", false, "Reduce ram usage by parsing BAM files sequentially")
	cpuprofile      = flag.String("cpu-prof", "", "Write cpu profile to `file` (used for benchmarking)")
	help            = flag.Bool("help", false, "Display help")
	mutex           = &sync.Mutex{}
	wg              sync.WaitGroup
)

func main() {
	flag.Parse()
	if *help {
		flag.Usage()
		os.Exit(0)
	}

	// Initialize CPU profiling
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal("could not create CPU profile: ", err)
		}
		if err := pprof.StartCPUProfile(f); err != nil {
			log.Fatal("could not start CPU profile: ", err)
		}
		defer pprof.StopCPUProfile()
	}

	// Validate input parameters
	if *fastaFile == "" {
		log.Fatal("Error: FASTA input file parameter is required.")
	}
	if *jsonOutputFile == "" {
		log.Fatal("Error: JSON output file parameter is required.")
	}
	var sampleNameList, bamFileNames, rpkmFileNames []string
	if *sampleBAMFiles != "" || *sampleRPKMFiles != "" {
		sampleNameList = strings.Split(*sampleNames, ",")
	}
	if *sampleBAMFiles != "" {
		if *sampleNames == "" {
		}
		bamFileNames = strings.Split(*sampleBAMFiles, ",")
		if len(sampleNameList) != len(bamFileNames) {
			log.Fatal("Error: Sample names / BAM files arguments differ in length.")
		}
	}
	if *sampleRPKMFiles != "" {
		if *sampleNames == "" {
			log.Fatal("Error: Sample names required for RPKM file parsing.")
		}
		rpkmFileNames = strings.Split(*sampleRPKMFiles, ",")
		if len(sampleNameList) != len(rpkmFileNames) {
			log.Fatal("Error: Sample names / RPKM files arguments differ in length.")
		}
	}

	start := time.Now()

	contigs := make(Contigs)

	// FASTA
	log.Print("Parsing contigs FASTA file ...")
	contigIDs := parseContigsFastaFile(*fastaFile, &contigs)

	// BAM
	if !*skipCoverage && *sampleBAMFiles != "" {
		numCPU := runtime.NumCPU()
		var threadsPerBAMFile int
		if *reduceRAMUsage {
			threadsPerBAMFile = numCPU + 2
			log.Println("sequentially in reduced RAM usage mode ...")
		} else {
			threadsPerBAMFile = int(math.Ceil(float64(numCPU)/float64(len(bamFileNames)))) + 1
		}
		bamFileNames = strings.Split(*sampleBAMFiles, ",")
		log.Printf("Parsing %d BAM files using %d threads per BAM file ...", len(bamFileNames), threadsPerBAMFile)
		for sampleIndex, sampleName := range sampleNameList {
			if *reduceRAMUsage {
				parseBAMFile(sampleName, bamFileNames[sampleIndex], &contigs, threadsPerBAMFile)
			} else {
				wg.Add(1)
				go parseBAMFile(sampleName, bamFileNames[sampleIndex], &contigs, threadsPerBAMFile)
			}
		}
		if !*reduceRAMUsage {
			wg.Wait()
		}
	}

	// RPKM
	if *sampleRPKMFiles != "" {
		log.Print("Parsing RPKM files ...")
		for sampleIndex, sampleName := range sampleNameList {
			parseRPKMFile(sampleName, rpkmFileNames[sampleIndex], &contigs)
		}
	}

	// VirSorter
	if *virSorterCSV != "" {
		log.Print("Parsing VirSorter file ...")
		parseVirSorterCSV(*virSorterCSV, &contigs)
	}

	// JSON
	log.Print("Writing JSON output file ...")
	objCount := writeJson(&contigs, contigIDs, *jsonOutputFile)
	log.Printf("%d JSON objects written to %q.", objCount, *jsonOutputFile)

	elapsed := time.Since(start)
	log.Printf("Total time: %s", elapsed)
}

func headerToID(header string) string {
	headerFields := strings.SplitN(header, " ", 2)
	if len(headerFields) < 2 || len(headerFields[0]) == 0 {
		log.Fatalf("Error while parsing header line: %q", header)
	}
	return headerFields[0]
}

func parseContigsFastaFile(path string, contigs *Contigs) []string {
	reader, err := fastx.NewDefaultReader(path)
	seq.ValidateSeq = false
	if err != nil {
		log.Fatalf("error reading FASTA: %v\n", err)
	}
	defer reader.Close()
	var contigIDs []string
	for {
		fastaRec, err := reader.Read()
		if err != nil {
			if err == io.EOF {
				break
			}
			if err != nil {
				log.Fatalf("error reading FASTA: %v\n", err)
			}
			break
		}
		(*contigs)[string(fastaRec.ID)] = &Contig{
			ID:         string(fastaRec.ID),
			Nucleotide: string(fastaRec.Seq.Seq),
			Length:     fastaRec.Seq.Length(),
		}
		contigIDs = append(contigIDs, string(fastaRec.ID))
	}
	return contigIDs
}

func openBAMFile(path string, threads int) (*os.File, *bam.Reader) {
	var r io.Reader
	f, err := os.Open(path)
	if err != nil {
		log.Fatalf("Could not open file %q\n", err)
	}
	ok, err := bgzf.HasEOF(f)
	if err != nil {
		log.Fatalf("Could not open file %q\n", err)
	}
	if !ok {
		log.Printf("File %q has no BGZF magic block: may be truncated\n", path)
	}
	r = f

	b, err := bam.NewReader(r, threads)
	if err != nil {
		log.Fatalf("Could not read BAM: %q\n", err)
	}
	b.SetCache(cache.NewLRU(6000))
	return f, b
}

func parseBAMFile(sampleName string, path string, contigs *Contigs, threads int) {
	if !*reduceRAMUsage {
		defer wg.Done()
	}
	f, b := openBAMFile(path, threads)
	defer f.Close()
	defer b.Close()
	b.Omit(bam.AllVariableLengthData)

	var lastSeenContig *sam.Reference
	var cov []int
	for {
		bamRec, err := b.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatalf("error reading BAM: %v\n", err)
		}
		if lastSeenContig == nil {
			lastSeenContig = bamRec.Ref
			cov = make([]int, bamRec.Ref.Len())
		}
		if bamRec.Ref != lastSeenContig {
			// wrap up current contig and save results
			contigID := headerToID(lastSeenContig.Name())
			c := Coverage{Name: sampleName, Values: calcCovAvg(&cov)}
			mutex.Lock()
			(*contigs)[contigID].Coverage = append((*contigs)[contigID].Coverage, &c)
			mutex.Unlock()
			// stop parser loop at the first unmapped sequence
			if bamRec.Ref.Name() == "*" {
				break
			}

			// prepare for mappings of next contig
			lastSeenContig = bamRec.Ref
			cov = make([]int, bamRec.Ref.Len())
		}
		if bamRec.Ref == lastSeenContig {
			pileupRec(bamRec, &cov)
		}
	}
}

func pileupRec(rec *sam.Record, cov *[]int) {
	recEnd := rec.End()
	for i := rec.Start(); i < recEnd; i++ {
		(*cov)[i]++
	}
}

func calcCovAvg(cov *[]int) []int {
	covAvg := make([]int, 0, len(*cov)/CovAvgSpan+1)
	for i := 0; i < len(*cov)-CovAvgSpan; i += CovAvgSpan {
		tmp := 0
		for j := i; j < i+CovAvgSpan; j++ {
			tmp += (*cov)[j]
		}
		covAvg = append(covAvg, tmp/CovAvgSpan)
	}
	return covAvg
}

func parseRPKMFile(sampleName string, path string, contigs *Contigs) {
	tsvGz, err := os.Open(path)
	if err != nil {
		log.Fatalf("Could not open file: %q\n", err)
	}
	defer tsvGz.Close()
	tsv, err := gzip.NewReader(tsvGz)
	if err != nil {
		log.Fatalf("Could not open gzipped file: %q\n", err)
	}
	defer tsv.Close()

	// skip header rows
	scanner := bufio.NewScanner(tsv)
	scanner.Scan()
	scanner.Scan()
	scanner.Scan()
	scanner.Scan()
	scanner.Scan()
	tsvContents := new(bytes.Buffer)
	for scanner.Scan() {
		tsvContents.WriteString(scanner.Text())
		tsvContents.WriteString("\n")
	}

	tsvReader := csv.NewReader(tsvContents)
	tsvReader.Comma = '\t'
	for {
		record, err := tsvReader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatalf("Error parsing RPKM file: %q\n", err)
		}
		contigID := headerToID(record[0])
		val, err := strconv.ParseFloat(record[5], 64)
		if err != nil {
			log.Fatalf("Error parsing RPKM value: %q\n", err)
		}
		r := Rpkm{Name: sampleName, Value: val}
		(*contigs)[contigID].Rpkm = append((*contigs)[contigID].Rpkm, &r)
	}
}

func parseVirSorterCSV(path string, contigs *Contigs) {
	vsFile, err := os.Open(path)
	if err != nil {
		log.Fatalf("Could not open file %q: %q\n", path, err)
	}
	defer vsFile.Close()
	var phageConfidence, prophageConfidence, state string
	scanner := bufio.NewScanner(vsFile)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		switch line {
		case "## 1 - Complete phage contigs - category 1 (sure)":
			phageConfidence = "sure"
			state = "phage"
		case "## 2 - Complete phage contigs - category 2 (somewhat sure)":
			phageConfidence = "somewhat sure"
			state = "phage"
		case "## 3 - Complete phage contigs - category 3 (not so sure)":
			phageConfidence = "not so sure"
			state = "phage"
		case "## 4 - Prophages - category 1 (sure)":
			prophageConfidence = "sure"
			state = "prophage"
		case "## 5 - Prophages - category 2 (somewhat sure)":
			prophageConfidence = "somewhat sure"
			state = "prophage"
		case "## 6 - Prophages - category 3 (not so sure)":
			prophageConfidence = "not so sure"
			state = "prophage"
		default:
			if !strings.HasPrefix(line, "#") {
				virsorter := &Virsorter{}
				contigID := strings.SplitN(strings.TrimPrefix(strings.SplitN(line, ",", 2)[0], "VIRSorter_"), "_flag=", 2)[0]
				if len(contigID) == 0 || strings.ContainsAny(contigID, "=") {
					log.Fatalf("VirSorter contig identifiers are being parsed incorrectly. "+
						"Original ID is %q, parsed contigID is %q", strings.SplitN(line, ",", 2)[0], contigID)
				}
				switch state {
				case "phage":
					virsorter.Phage = phageConfidence
				case "prophage":
					virsorter.Prophage = prophageConfidence
				}
				(*contigs)[contigID].Virsorter = virsorter
			}
		}

	}
}

func writeJson(contigs *Contigs, contigIDs []string, path string) int {
	out, err := os.Create(path)
	if err != nil {
		log.Fatalf("Could not open file for writing: %q\n", err)
	}
	defer out.Close()
	var i int
	for _, k := range contigIDs {
		_, err := fmt.Fprintf(out, "{\"index\": {\"_id\": \"%s\"}}\n", k)
		if err != nil {
			log.Fatalf("Could not write to file: %q\n", err)
		}
		obj, err := json.Marshal((*contigs)[k])
		if err != nil {
			log.Fatalf("Could not generate JSON object for key %s:", k)
		}
		_, err = fmt.Fprintf(out, "%s\n", string(obj))
		if err != nil {
			log.Fatalf("Could not write to file: %q\n", err)
		}
		i++
	}
	return i
}
